### PR TITLE
winpr/file refactoring and regular file support

### PR DESCRIPTION
**Pull request by bmiklautz**
* refactor handle creator
* split out client side named pipe handling from generic file handling
* add initial support for regular files.

### CODE REVIEW

File: winpr/libwinpr/file/generic.c [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+}
+
+BOOL CreateDirectoryW(LPCWSTR lpPathName, LPSECURITY_ATTRIBUTES lpSecurityAttributes)
+{
+	return FALSE;
+}
+
+BOOL RemoveDirectoryA(LPCSTR lpPathName)
+{
+	if (!rmdir(lpPathName))
```
Comment (hardening) [inline winpr/libwinpr/file/generic.c]: return rmdir(lpPathName) == 0 ?

**Comment by hardening** [comment]
I'm not a WIN32 specialist but AFAICT it LGTM

File: winpr/libwinpr/file/generic.c (hunk 1 above)
Comment (bmiklautz) [inline winpr/libwinpr/file/generic.c]: @hardening this was just split moved out from the old file.c I didn't review the code. But your suggestion makes sense.

**Comment by bmiklautz** [comment]
@hardening the one thing need to test on Windows is what happens if you close the handle for stdin/stderr/stdout. And we CreateFile is also missing but something for another pull request then.

**Comment by mfleisz** [comment]
If you close one of the std* handles on Windows then the file descriptor will simply be closed and rendered invalid. You would have to set a new handle (i.e. using SetStdHandle) to reenable the descriptor.
See also http://blogs.msdn.com/b/oldnewthing/archive/2013/03/07/10399690.aspx

**Comment by mfleisz** [comment]
(Can't edit with stupid github app)
So I think the behavior will be the same as on the other platforms right?

**Comment by bmiklautz** [comment]
@mfleisz thanks for pointing that out. On Linux/UNIX closing the file descriptor handles it invalid but the next open call can be re-use the number. For example if a close(2) is done the fd returned by the next open (or similar) might be 2.

**Comment by mfleisz** [comment]
@bmiklautz So on Linux - if you close stdin (usually value 0 iirc) and you open a file then it will get a descriptor id of 0 --> so stdin automatically (without setting it directly or via an API) points to that file? That behavior is different from Windows then ... Windows won't set the stdin handle automatically. the std handles are just global vars that can only be accessed via the Set/GetStdHandle APIs - and if you close that handle it will also make that global variable (used by all the IO functions) useless.

File: winpr/include/winpr/file.h [hunk 2]
```diff
@@ -329,7 +332,9 @@ typedef struct _HANDLE_CREATOR
 	pcCreateFileA CreateFileA;
 } HANDLE_CREATOR, *PHANDLE_CREATOR, *LPHANDLE_CREATOR;
 
-BOOL RegisterHandleCreator(PHANDLE_CREATOR pHandleCreator);
+WINPR_API HANDLE GetStdHandle(DWORD nStdHandle);
+WINPR_API BOOL SetStdHandle(DWORD nStdHandle, HANDLE hHandle);
+WINPR_API BOOL SetStdHandleEx(DWORD dwStdHandle, HANDLE hNewHandle, HANDLE* phOldHandle);
```
Comment (MartinHaimberger) [inline winpr/include/winpr/file.h]: this exports should be in the extern "C" { } section ....

**Comment by bmiklautz** [comment]
@mfleisz correct. The standard handle are only the first file handles the process has thats why you usually try to keep them "occupied". What we can do is if the handle is closed to redirect the corresponding fd to /dev/null and if it is set again we re-open it. I think that would come the closest to the behavior on windows. 
I'll create a new pull request for that then.
Thank for your hints!

**Comment by mfleisz** [comment]
@bmiklautz Sounds good :+1:

**Comment by MartinHaimberger** [comment]
+1 LGTM

**Pull request merged**